lisa/tools/virtualclient.py

```diff
@@ -0,0 +1,402 @@
+            metric_messages: List[VCMetrics] = []
+            for row in reader:
+                metric = VCMetrics(
+                    time_stamp=self._parse_timestamp(row["Timestamp"]),
+                    experiment_id=row["ExperimentId"],
+                    client_id=row["ClientId"],
+                    profile=row["Profile"],
+                    profile_name=row["ProfileName"],
+                    tool_name=row["ToolName"],
+                    scenario_name=row["ScenarioName"],
+                    scenario_start_time=self._parse_timestamp(row["ScenarioStartTime"]),
+                    scenario_end_time=self._parse_timestamp(row["ScenarioEndTime"]),
+                    metric_categorization=row["MetricCategorization"],
+                    metric_name=row["MetricName"],
+                    metric_value=Decimal(row["MetricValue"]),
+                    metric_unit=row["MetricUnit"],
+                    metric_description=row["MetricDescription"],
+                    metric_relativity=row["MetricRelativity"],
+                    execution_system=row["ExecutionSystem"],
+                    operating_system_platform=row["OperatingSystemPlatform"],
+                    operation_id=row["OperationId"],
+                    operation_parent_id=row["OperationParentId"],
+                    app_host=row["AppHost"],
+                    app_name=row["AppName"],
+                    app_version=row["AppVersion"],
+                    app_telemetry_version=row["AppTelemetryVersion"],
+                    tags=row["Tags"],
+                )
+                metric_messages.append(metric)
+        for metri_messages in metric_messages:
+            notifier.notify(metri_messages)
+
+    def _parse_timestamp(self, timestamp_str: str) -> datetime:
+        # Strip 'Z' and limit microseconds to 6 digits
+        timestamp_str = timestamp_str.rstrip("Z")
+        # Split timestamp and microseconds
+        if "." in timestamp_str:
+            main_time, microseconds = timestamp_str.split(".")
+            # Truncate microseconds to 6 digits
+            microseconds = microseconds[:6]
+            timestamp_str = f"{main_time}.{microseconds}"
+        return datetime.strptime(timestamp_str, "%Y-%m-%dT%H:%M:%S.%f")
+
+
+class VcRunner:
+    # [07/08/2024 13:58:15] Exit Code: 0\r\n
+    exit_code_pattern = re.compile(r"([\w\W]*?)Exit Code: (?P<exit_code>\d+)")
+
+    def __init__(self, environment: "Environment", roles: List[str]):
+        if len(roles) > len(environment.nodes):
+            raise ValueError(
+                "More roles specified than available nodes in the environment"
+            )
+
+        self._targets: List[VcTargetInfo] = []
+        nodes = [cast("RemoteNode", environment.nodes[i]) for i in range(len(roles))]
+
+        for role, node in zip(roles, nodes):
+            self._targets.append(VcTargetInfo(role=role, node=node))
+
+        self._virtual_clients: List[Any] = run_in_parallel(
```

Comment:
Put virtual clients into the `VcTargetInfo`, so it's saved with its role/node together, and easy to loop.